Framework/DataHandling/src/LoadNGEM.cpp

```diff
@@ -0,0 +1,473 @@
+ * @param minEventsReq The number of events required to be a good frame.
+ * @param maxEventsReq The max events allowed to be a good frame.
+ * @param frameEventCounts A vector of the number of events in each good frame.
+ * @param histograms The main set of histograms for the data so far.
+ * @param histogramsInFrame The set of histograms for the current frame.
+ */
+void LoadNGEM::addFrameToOutputWorkspace(
+    int &rawFrames, int &goodFrames, const int &eventCountInFrame,
+    const int &minEventsReq, const int &maxEventsReq,
+    MantidVec &frameEventCounts,
+    std::vector<DataObjects::EventList> &histograms,
+    std::vector<DataObjects::EventList> &histogramsInFrame) {
+  ++rawFrames;
+  if (eventCountInFrame >= minEventsReq && eventCountInFrame <= maxEventsReq) {
+    // Add number of event counts to workspace.
+    frameEventCounts.emplace_back(eventCountInFrame);
+    ++goodFrames;
+
+#pragma omp parallel for
+    // Add histograms that match parameters to workspace
+    for (auto i = 0; i < NUM_OF_SPECTRA; ++i) {
+      if (histogramsInFrame[i].getNumberEvents() > 0) {
+        histograms[i] += histogramsInFrame[i];
+        histogramsInFrame[i].clear();
+      }
+    }
+  }
+}
+
+/**
+ * @brief Report the progress of the loader through the current file.
+ *
+ * @param numProcessedEvents The number of events processed so far.
+ * @param eventCountInFrame The number of events in the current frame.
+ * @param totalNumEvents The total number of events in the file.
+ * @param totalFilePaths The total number of file paths to process.
+ * @param fileCount The number of files processed.
+ * @return true If user has cancelled the load.
+ * @return false If the user has not cancelled.
+ */
+bool LoadNGEM::reportProgressAndCheckCancel(size_t &numProcessedEvents,
+                                            int &eventCountInFrame,
+                                            const size_t &totalNumEvents,
+                                            const size_t &totalFilePaths,
+                                            const int &fileCount) {
+  numProcessedEvents += eventCountInFrame;
+  std::string message(std::to_string(fileCount) + "/" +
+                      std::to_string(totalFilePaths));
+  progress(double(numProcessedEvents) / double(totalNumEvents) / 1.11111,
+           message);
+  eventCountInFrame = 0;
+  // Check for cancel flag.
+  return this->getCancel();
+}
+
+/**
+ * @brief Creates an event workspace and fills it with the data.
+ *
+ * @param maxToF The largest ToF seen so far.
+ * @param binWidth The width of each bin.
+ * @param histograms The main histogram event data.
+ * @param dataWorkspace The workspace to add the data to.
+ */
+void LoadNGEM::createEventWorkspace(
+    const int &maxToF, const double &binWidth,
+    std::vector<DataObjects::EventList> &histograms,
+    DataObjects::EventWorkspace_sptr &dataWorkspace) {
+  std::vector<double> xAxis;
```

Comment:
If we know the size of a vector upfront it is a good idea to call `.reserve` on it to avoid repeated memory allocation and ended up with a vector that has over allocated space.